### PR TITLE
Draggable overlays

### PR DESCRIPTION
**Pull request by howard-stearns**
1. Overlay2D::setX()/setY() now do that. They used to stretch the overlay by moving the upper left corner while keeping the lower right in place. Now the overlay is moved.

2. Draggable 2D overlays, per https://app.asana.com/0/32622044445063/36841246548538

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7602c93652dca209bbe3097b5fad24b9fecfb910 available at: https://gist.github.com/c1539962b76dd2c3e297<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5089.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5089.dmg

File: interface/src/ui/overlays/Overlays.cpp [hunk 1]
```diff
@@ -293,6 +293,45 @@ unsigned int Overlays::getOverlayAtPoint(const glm::vec2& point) {
...
+    glm::vec2 mousePoint(app->getMouseX(), app->getMouseY());
+    unsigned int id = getOverlayAtPoint(mousePoint);
+    if (!id) {
+        return;
+    }
+    Overlay::Pointer candidateOverlay = _overlaysHUD[id];
+    if (candidateOverlay->is3D()) {
+        return;
+    }
+    _dragged2dOverlay = static_cast<Overlay2D*>(candidateOverlay.get());;
```
Comment (ZappoMan) [inline interface/src/ui/overlays/Overlays.cpp]: dies ist verboten!

Do not make a non-shared pointer of a shared pointer... that's bad business... instead use:

```
std::shared_ptr<Overlay2D> _dragged2dOverlay;
_dragged2dOverlay  = std::dynamic_pointer_cast<Overlay2D>(candidateOverlay);
```

File: interface/src/ui/overlays/Overlays.cpp [hunk 2]
```diff
@@ -293,6 +293,45 @@ unsigned int Overlays::getOverlayAtPoint(const glm::vec2& point) {
...
+    _dragged2dOverlayOffsetX = _dragged2dOverlay->getX() - app->getMouseX();
+    _dragged2dOverlayOffsetY = _dragged2dOverlay->getY() - app->getMouseY();
+}
+
+void Overlays::mouseMoveEvent(QMouseEvent* event, unsigned int deviceID) {
+    if (!_dragged2dOverlay ||
+        // We could have a separate mouseReleaseEvent() clear the overlay, but this will be more reliable.
+        // Note also that QMouseEvent::button() (singular) is null for moves. See qt doc.
+        (event->buttons() == Qt::NoButton)) {
```
Comment (ZappoMan) [inline interface/src/ui/overlays/Overlays.cpp]: this indenting makes it hard to read... I'd suggest putting the entire if clause on one line and including your comment above the if

File: interface/src/ui/overlays/Overlays.h [hunk 3]
```diff
@@ -97,6 +100,12 @@ public slots:
     QReadWriteLock _lock;
     QReadWriteLock _deleteLock;
     QScriptEngine* _scriptEngine;
+
+    // Support for dragging overlays.
+    // Currently only supports dragging one 2D overlay at a time. (No multi-touch.)
+    Overlay2D* _dragged2dOverlay;
```
Comment (ZappoMan) [inline interface/src/ui/overlays/Overlays.h]:
```
std::shared_ptr<Overlay2D> _dragged2dOverlay;
```

**Comment by ZappoMan** [comment]
This appears to make all 2D overlays draggable... I'm not sure I would want that behavior for my 2D overlays... what's to prevent someone from moving all the overlays around and having the UI totally borked?

File: interface/src/ui/overlays/Overlays.cpp (hunk 1 above)
Comment (howard-stearns) [inline interface/src/ui/overlays/Overlays.cpp]: Entschuldigen Sie. Vielen Dank.
Will fix. (and indentation)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6dc1e9bc6176fc27842e51c38b1da30061bccb19 available at: https://gist.github.com/c1539962b76dd2c3e297<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5089.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5089.dmg

**Comment by howard-stearns** [comment]
Putting this aside for now, in favor of doing things through Javascript.

**Pull request closed without merging**